### PR TITLE
Switch to new incremental compiler

### PR DESCRIPTION
**Pull request by mraleph**
These are changes that will need to be applied to frontend server once IKG is removed and replaced with new Fasta centric incremental compiler.

### CODE REVIEW

**Comment by mraleph** [comment]
This should be merge when the Dart SDK part of it is landed together with the Dart roll.

File: frontend_server/lib/server.dart [hunk 1]
```diff
@@ -298,9 +277,9 @@ Future<int> starter(
           boundaryKey = string.substring(RECOMPILE_INSTRUCTION_SPACE.length);
           state = _State.RECOMPILE_LIST;
         } else if (string == 'accept') {
-          compiler.acceptLastDelta();
+          // NOP
```
Comment (aam) [inline frontend_server/lib/server.dart]: Perhaps add a TODO explaining the change to track removal of these commands from flutter/flutter too?

Comment (mraleph) [reply frontend_server/lib/server.dart]: We might need to restore this. I am iterating on this with Peter.

File: frontend_server/lib/server.dart [hunk 2]
```diff
@@ -96,14 +96,6 @@ abstract class CompilerInterface {
   /// taking into account some changed(invalidated) sources.
   Future<Null> recompileDelta();
 
-  /// Accept results of previous compilation so that next recompilation cycle
-  /// won't recompile sources that were previously reported as changed.
-  void acceptLastDelta();
```
Comment (aam) [inline frontend_server/lib/server.dart]: With these gonedDoes the `resetIncrementalCompiler` below still makes sense?

Comment (mraleph) [reply frontend_server/lib/server.dart]: I have implemented `resetIncrementalCompiler` now.

**Comment by mraleph** [comment]
@aam could you take another look?

File: frontend_server/lib/server.dart [hunk 3]
```diff
@@ -192,35 +184,28 @@ class _FrontendCompiler implements CompilerInterface {
...
-
-  @override
-  void rejectLastDelta() {
-    _generator.rejectLastDelta();
-  }
-
   @override
   void invalidate(Uri uri) {
-    _generator.invalidate(uri);
+    _generator?.invalidate(uri);
```
Comment (aam) [inline frontend_server/lib/server.dart]: Why is this change needed? We don't expect `invalidate` being called without `_generator` in place, do we?

Comment (mraleph) [reply frontend_server/lib/server.dart]: Oops. Left over from the time when I was just nulling out _generator. Reverting

File: frontend_server/lib/server.dart [hunk 4]
```diff
@@ -309,9 +291,9 @@ Future<int> starter(
           boundaryKey = string.substring(RECOMPILE_INSTRUCTION_SPACE.length);
           state = _State.RECOMPILE_LIST;
         } else if (string == 'accept') {
-          compiler.acceptLastDelta();
+          // TODO(vegorov) remove this command from the protocol
         } else if (string == 'reject') {
-          compiler.rejectLastDelta();
+          // TODO(vegorov) remove this command from the protocol
```
Comment (aam) [inline frontend_server/lib/server.dart]: Curious, how does incremental compiler work without `rejectLastDelta`?
How does incremental compiler know when to included rejected changes(i.e. changes that VM could not load) into next incremental kernel file?

Comment (mraleph) [reply frontend_server/lib/server.dart]: I guess now it this does not necessarily work. Are there any tests or design documentation for this behavior? I would like to learn what sort of changes can be rejected by the VM and how unrolling front-end state fixes it.

/cc @peter-ahe-google

Comment (peter-ahe-google) [reply frontend_server/lib/server.dart]: For now, please create a new instance. I don't understand why the incremental compiler needs to keep track of what VM doesn't like, but I'll be happy to help you move this kind of logic to package/vm.

Comment (mraleph) [reply frontend_server/lib/server.dart]: Implemented reset on reject.

Comment (aam) [reply frontend_server/lib/server.dart]: The best documentation I can think of is this: https://github.com/flutter/engine/tree/master/frontend_server#acceptreject

I'm thinking about class->enum kind of changes that current VM hot reload won't accept. Not sure what else VM hot reload would not be happy with. @rmacnak-google can comment.

In previous IKG concept was that frontend needs to include files that resulted in rejected hot reload request so that VM sees complete incremental picture since last time incremental file was 
accepted.

**Comment by aam** [comment]
LGTM considering that this unblocks dart/flutter roll.

File: frontend_server/lib/server.dart [hunk 5]
```diff
@@ -192,35 +184,28 @@ class _FrontendCompiler implements CompilerInterface {
...
-  @override
-  void acceptLastDelta() {
-    _generator.acceptLastDelta();
-  }
-
-  @override
-  void rejectLastDelta() {
-    _generator.rejectLastDelta();
-  }
-
```
Comment (a-siva) [inline frontend_server/lib/server.dart]: I am not sure why these functions should be deleted and not just made nops temporarily with a TODO that refers to the issue that needs fixing. I have created a dart issue here https://github.com/dart-lang/sdk/issues/31919

File: frontend_server/lib/server.dart [hunk 6]
```diff
@@ -298,9 +291,11 @@ Future<int> starter(
           boundaryKey = string.substring(RECOMPILE_INSTRUCTION_SPACE.length);
           state = _State.RECOMPILE_LIST;
         } else if (string == 'accept') {
-          compiler.acceptLastDelta();
+          // TODO(vegorov) remove this command from the protocol
```
Comment (a-siva) [inline frontend_server/lib/server.dart]: I don't understand why this comment says remove this from the protocol, if we don't delete the methods acceptLastDelta, reject etc. and instead just made them nops this code would not have to change,

File: frontend_server/lib/server.dart [hunk 7]
```diff
@@ -298,9 +291,11 @@ Future<int> starter(
...
         } else if (string == 'reject') {
-          compiler.rejectLastDelta();
+          // TODO(vegorov) evaluate conditions under which deltas can be
+          // rejected by the VM.
+          compiler.resetIncrementalCompiler();
```
Comment (a-siva) [inline frontend_server/lib/server.dart]: Ditto comment about this change.

File: frontend_server/lib/server.dart (hunk 4 above)
Comment (rmacnak-google) [reply frontend_server/lib/server.dart]: Peter, I don't think the incremental compiler needs to know what kind of changes the VM doesn't support. It does need to know what is currently loaded into the VM to produce a correct delta, and needs to be aware that the most recent delta it produced might not have been loaded into the VM (due to an IO error or a kind of change the VM doesn't know how to handle), but it need not try to eagerly detect the extra restrictions the VM adds.

Slava, see callers of AddReasonForCancelling for the kinds of changes the VM does not support. This delta tracking business should allow someone to make a kind of change the VM does not support, have the change be rejected, then make a different kind of change the VM does support and have the second change be properly loaded as if the first was never attempted.

Comment (mraleph) [reply frontend_server/lib/server.dart]: @rmacnak-google makes total sense. Thanks for clarifications. 

I think it would be great to ensure that we have coverage for these features somewhere in the test suite (does not matter if in Flutter/Flutter Engine or SDK repos). 

We are gonna implement rejection handling in the separate CL on top of public incremental compilation APIs.

File: frontend_server/lib/server.dart (hunk 5 above)
Comment (mraleph) [reply frontend_server/lib/server.dart]: We will address this issue after this CL lands.

File: frontend_server/lib/server.dart (hunk 4 above)
Comment (peter-ahe-google) [reply frontend_server/lib/server.dart]: I'll be happy to help getting rejection working on top of the front-end API.

**Comment by Hixie** [comment]
What's the status of this PR?

**Comment by mraleph** [comment]
Merged as part of 2b398ee2279c00c9b359e7d4bd06c27a4e657fc2.
